core/src/test/java/com/graphhopper/GraphHopperAPITest.java

```diff
@@ -115,4 +124,69 @@ public void testNoLoad() {
             assertTrue(ex.getMessage(), ex.getMessage().startsWith("Do a successful call to load or importOrLoad before routing"));
         }
     }
+
+    @Test
+    public void testConcurrentGraphChange() throws InterruptedException {
+        final GraphHopperStorage graph = new GraphBuilder(encodingManager).create();
+        initGraph(graph);
+        graph.edge(1, 2, 10, true);
+
+        final AtomicInteger checkPointCounter = new AtomicInteger(0);
+        final GraphHopper graphHopper = new GraphHopper() {
+            @Override
+            protected ChangeGraphHelper createChangeGraphHelper(Graph graph, LocationIndex locationIndex) {
+                return new ChangeGraphHelper(graph, locationIndex) {
+                    @Override
+                    public long applyChanges(EncodingManager em, Collection<JsonFeature> features) {
+                        // force sleep inside the lock
+                        try {
+                            Thread.sleep(1000);
+                        } catch (InterruptedException e) {
+                            throw new RuntimeException(e);
+                        }
+                        checkPointCounter.incrementAndGet();
+                        return super.applyChanges(em, features);
+                    }
+                };
+            }
+        }.setStoreOnFlush(false).setEncodingManager(encodingManager).setCHEnabled(false).
+                loadGraph(graph);
+
+        GHResponse rsp = graphHopper.route(new GHRequest(42, 10.4, 42, 10));
+        assertFalse(rsp.toString(), rsp.hasErrors());
+        assertEquals(1800, rsp.getBest().getTime());
+
+        final List<JsonFeature> list = new ArrayList<>();
+        Map<String, Object> properties = new HashMap<>();
+        properties.put("speed", 5);
+
+        list.add(new JsonFeature("1", "bbox",
+                new BBox(10.399, 10.4, 42.0, 42.001),
+                null, properties));
+
+        Thread thread1 = new Thread() {
+            @Override
+            public void run() {
```

Comment:
Try inserting
```
                    Thread.sleep(1);
```
here, at the beginning of thread1.
The test will fail, because we will not have even started locking the graph yet when the second routing request comes, and it will (correctly) see the old version.
That means that this test isn't really testing a stable situation at the moment.
I found this because I asked myself why we even need thread2: We can just as well send the second routing request on the main thread. But then the test fails, because opening thread2 apparently takes 1 microsecond, and when we are faster (_by not doing that_), we have the same situation as above.

So at the moment, you are only using thread 2 to fix a race condition.

Not a full solution, but: The second routing request should not be sent before we have reason to believe that thread1 has already had enough time to lock the graph.